### PR TITLE
Optimize primitive Array(e1, ..., en)

### PR DESCRIPTION
**Pull request by retronym**
Expands an existing optimization for reference arrays to
apply to primitives, as well.

Fixes one aspect of SI-6247.

Review by @paulp. The only difference from last time you saw this is `copyAttrs`.

### CODE REVIEW

**Comment by jsuereth** [comment]
BTW - We figured out the build bot issues.  Github API restrictions.  Someone was using giter8 under the same IP as the bot, so github was preventing comments from coming through.  YAY FOR RATE LIMITS.

File: src/compiler/scala/tools/nsc/transform/CleanUp.scala [hunk 1]
```diff
@@ -624,11 +624,11 @@ abstract class CleanUp extends Transform with ast.TreeDSL {
...
       case Apply(appMeth, List(Apply(wrapRefArrayMeth, List(arg @ StripCast(ArrayValue(_, _)))), _))
-      if (wrapRefArrayMeth.symbol == Predef_wrapRefArray &&
-          appMeth.symbol == ArrayModule_overloadedApply.suchThat {
-            _.tpe.resultType.dealias.typeSymbol == ObjectClass  // [T: ClassTag](xs: T*): Array[T] post erasure
-          }) =>
+      if wrapRefArrayMeth.symbol == Predef_wrapRefArray && appMeth.symbol == ArrayModule_genericApply =>
         super.transform(arg)
+      case Apply(appMeth, List(elem0, Apply(wrapArrayMeth, List(rest @ ArrayValue(elemtpt, _)))))
+      if wrapArrayMeth.symbol == Predef_wrapArray(elemtpt.tpe) && appMeth.symbol == ArrayModule_apply(elemtpt.tpe) =>
+        super.transform(rest.copy(elems = elem0 :: rest.elems).copyAttrs(rest))
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/transform/CleanUp.scala]: I do not think it is wise for us to start using the case class copy method on Trees without a) wide agreement this is a good idea and b) somehow making it do the copyAttrs part automatically, which AFAIK we can't yet do. In the absence of both those, I think this should be

```
treeCopy.ArrayValue(rest, elemtpt, elem0 :: rest.elems)
```

like it is elsewhere.

Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/CleanUp.scala]: Yes, you're right. I looked for that method on `TreeCopier` but didn't find it; but I see it now. (I must have been up in user-land in `reflect.api`).

**Comment by retronym** [comment]
Is it just me or is Github relegating discussions to the "outdated diff" category incorrectly? I just added a commit, not a `-f` in sight.

**Comment by paulp** [comment]
Maybe an "outdated diff" is any bit of code which has changed since the comment was left.

**Comment by retronym** [comment]
Oh, that's actually pretty useful.

**Comment by som-snytt** [comment]
It might be different for you guys, but my reaction was, I didn't craft a response for paulp only to have it hidden behind an outdated diff fold.  The code may be outdated but the wit is eternal.  Or the wisdom, were there any.

**Comment by retronym** [comment]
Indeed. Now who can greasemonkey up a weathered filter over the diff itself, and a radiant glow around the prose?

**Comment by retronym** [comment]
PLS REBUILD ALL, y'all.

**Comment by retronym** [comment]
:cat: https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/665/console

**Comment by adriaanm** [comment]
bump @paulp

**Comment by paulp** [comment]
LGTM

**Pull request merged**